src/v/storage/log_manager.cc

```diff
@@ -127,31 +130,49 @@ ss::future<> log_manager::housekeeping() {
      * them as fast as we can since the majority of the time they will be in the
      * hotpath / (request-response)
      */
+    std::unique_ptr<model::ntp> current_key{};
     using bflags = log_housekeeping_meta::bitflags;
-    return ss::do_until(
-             [this] {
-                 auto it = find_next_non_compacted_log(_logs);
-                 return it == _logs.end();
-             },
-             [this, collection_threshold] {
-                 return ss::with_scheduling_group(
-                   _config.compaction_sg, [this, collection_threshold] {
-                       auto it = find_next_non_compacted_log(_logs);
-                       if (it == _logs.end()) {
-                           // must check again because between the stop
-                           // condition and this continuation we might have
-                           // removed the log
-                           return ss::now();
-                       }
-                       it->second.flags |= bflags::compacted;
-                       it->second.last_compaction = ss::lowres_clock::now();
-                       return it->second.handle.compact(compaction_config(
+    return ss::repeat([this,
+                       collection_threshold,
+                       current_key = std::move(current_key)]() mutable {
+               return ss::with_scheduling_group(
+                 _config.compaction_sg,
+                 [this,
+                  collection_threshold,
+                  current_key = std::move(current_key)]() mutable {
+                     typename logs_type::iterator it, start_it;
+
+                     if (!current_key) {
+                         current_key.reset(new model::ntp());
```

Comment:
you might consider switching to `std::optional<model::ntp>` to express your intent more explicitly where `current_key` is initially unset. i'm not seeing a reason it has to be a unique_ptr on the heap.